network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,387 @@
+
+        // create first handshake message  (-> e, es, s, ss)
+        let mut rng = rand::rngs::OsRng;
+        let mut first_message = [0u8; noise::handshake_init_msg_len(0)];
+        let initiator_state = self
+            .0
+            .initiate_connection(
+                &mut rng,
+                &prologue,
+                remote_public_key,
+                None,
+                &mut first_message,
+            )
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("noise: wrong input passed {}", e),
+                )
+            })?;
+
+        // write the first handshake message
+        socket.write_all(&first_message).await?;
+
+        // flush
+        socket.flush().await?;
+
+        // receive the server's response (<- e, ee, se)
+        let mut server_response = [0u8; noise::handshake_resp_msg_len(0)];
+        socket.read_exact(&mut server_response).await?;
+
+        // parse the server's response
+        // TODO: security logging here? (mimoo)
+        let (_, session) = self
+            .0
+            .finalize_connection(initiator_state, &server_response)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // finalize the connection
+        Ok(NoiseSession::new(socket, session))
+    }
+
+    pub async fn accept<TSocket>(
+        &self,
+        mut socket: TSocket,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<NoiseSession<TSocket>>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // receives prologue as the client timestamp in seconds
+        let mut prologue = [0u8; 8];
+        socket.read_exact(&mut prologue).await?;
+        let client_timestamp_u64 = u64::from_le_bytes(prologue);
+        let client_timestamp = time::Duration::from_secs(client_timestamp_u64);
+
+        // check the client timestamp
+        // TODO: security logging (mimoo)
+        let now = time::SystemTime::now()
+            .duration_since(time::UNIX_EPOCH)
+            .expect("system clock should work");
+        if client_timestamp > now
+            && client_timestamp - now > time::Duration::from_secs(MAX_FUTURE_TIMESTAMP)
+        {
+            println!("DEBUG: 2");
+            // if the client timestamp is too far in the future, abort
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!(
+                    "noise: client initiated connection with a timestamp too far in the future: {}",
+                    client_timestamp_u64
+                ),
+            ));
+        } else if now.checked_sub(client_timestamp).unwrap()
+            > time::Duration::from_secs(EXPIRATION_TIMESTAMP)
+        {
```

Comment:
maybe push these condition checks into their own lines above so we have nice clean if too_far_into_the future { ... } else too_old { ... }

Reply:
replaced with a match, I think it's clearer now